functions/autoload/__p9k_get_unique_path

```diff
@@ -16,24 +16,46 @@ setopt localoptions # extendedglob
 #   Given a directory path, determine the unique path.
 ##
 # @args
-#   $1 string The directory path to be truncated.
+#   $1 string Absolute directory path to be truncated.
+##
+# @info
+#   Takes paths like /abs/dir/path or ~/abs/home/dir/path but
+#   not a relative path like ./dir/path or dir/path
+#   The path "." will be ignores.
+#   Does not work well with characters that have to be escaped, e.g. " ", "*"
 ##
 # @function __p9k_get_unique_path() {}
-local trunc_path directory test_dir test_dir_length
+local trunc_path directory test_dir cur_path
 local -a matching
 local -a paths
-local cur_path='/'
+
 paths=(${(s:/:)1})
+if [[ "$1" == "." ]]; then
+  return 0
+elif [[ "$1" == "~"* ]]; then
+  # remove ~ from $paths and "move" $cur_paths to $HOME
+  paths=(${paths[@]:1})
+  cur_path="$HOME/"
+  trunc_path='~/'
+else
+  cur_path='/'
+  trunc_path='/'
+fi
+
 for directory in ${paths[@]}; do
+  # finds out how many characters you need for the $directory to be unique
   test_dir=''
   for (( i=0; i < ${#directory}; i++ )); do
     test_dir+="${directory:$i:1}"
-    matching=("$cur_path"/"$test_dir"*/)
+    matching=("$cur_path/$test_dir"*/)
```

Comment:
Yes. Currently, you try to find a unique name by building up a glob that is every time a character longer, as long you did not find exactly one match. This is a hit to the disk every time.
My hypothesis is now that you have all information necessary after the first glob to build up a unique directory.

Example:
In a folder structure like this:
```
.
├── asdf
├── asdf100
├── asdf101
├── asdf102
├── asdf2
├── asdf3
├── asdf4
├── asdf5
├── asdf6
├── asdf7
│   └── a
│       └── b
│           └── c
├── asdf8
├── asdf89
└── asdf99
```
we search the folder `asdf7` (Current code)
```zsh
declare -a matching
directory="asdf7"
test_dir=""
for (( i=0; i < ${#directory}; i++ )); do
  test_dir+="${directory:$i:1}"
  echo $test_dir;
  matching=("${PWD}/${test_dir}"*/);
  echo $matching
done
```

If we modify that loop a bit:
```zsh
declare -a matching
directory="asdf7"
test_dir=""
matching=("${PWD}/${directory[1]}"*/);
for (( i=0; i < ${#directory}; i++ )); do
  test_dir+="${directory:$i:1}"
  echo "Testing: $test_dir";
  pattern="${PWD}/${test_dir}*"
  matching=("${(M)matching:#${pattern}}")
  echo $matching
done
```
(not functional, but you get the idea).
This will need some measurings, but I think the second variant will be faster.